### PR TITLE
Fix urlFor() to work with optional parameters

### LINKED ISSUE #1289: Optional parameter in url
Slim doesn't support optional parameters as the underlying routing engine doesn't support them.

We need:
- Ability to specify optional parameters in the route path
- Ability to generate urls via `urlFor` for routes that have optional parameters.

This was originally reported as:

It would seem it currently isn't possible to add an optional paramater in the url if you are using urlFor to get the url via a named route.

``` php
$app->get('/test/{param1}[/{optional:[A-Za-z]}]')->setName('test');

$app->getContainer()['router']->urlFor('test', ['param1' => 'World']);
```

This will throw an error because urlFor method does not know how to handle optional regex. I saw another project using FastRoute and they handle it using this regex https://github.com/mrjgreen/phroute/blob/master/src/Phroute/RouteParser.php#L23-L29

### PR DESCRIPTION
**Pull request by akrabat**
Also, update composer.json to pick up the latest FastRoute until Nikita creates a new tag.

This PR fixes #1289.

### CODE REVIEW

**Comment by JoeBengalen** [comment]
Not sure if it is smart to depend on version `dev-master`. Why not specify version `0.5`?

**Comment by akrabat** [comment]
Version 0.5 doesn't have optional routes as it was tagged on 5th May 2015 and optional parameters was implemented on 18th June 2015.

**Comment by JoeBengalen** [comment]
Oh OK, thought it was the update from 0.4 to 0.5. Makes sense now ;) But it would be nice if for the final version of Slim we could depend on a specific version of fast-route.

**Comment by akrabat** [comment]
I agree. We'll ask @nikic nicely sometime after beta1 when it's had some more widespread testing :)

File: tests/RouterTest.php [hunk 1]
```diff
@@ -83,6 +83,30 @@ public function testUrlFor()
         );
     }
 
+    public function testUrlForWithOptionalParameters()
+    {
+        $methods = ['GET'];
+        $pattern = '/archive/{year}[/{month:[\d:{2}]}[/d/{day}]]';
```
Comment (nikic) [inline tests/RouterTest.php:89]: Is the `[\d:{2}]` here intentional?

Comment (akrabat) [inline tests/RouterTest.php:89]: Yes - I wanted to ensure that it didn't break anything

**Pull request merged**